libqtile/widget/decorations.py

```diff
@@ -0,0 +1,227 @@
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+import math
+from typing import Any, List, Tuple
+
+from libqtile.log_utils import logger
+from libqtile.widget import base
+
+
+class _Decoration(base.PaddingMixin):
+    """
+        Base decoration class. Should not be called by
+        configs directly.
+    """
+
+    defaults = [
+        ("padding", 0, "Default padding")
+    ]  # type: List[Tuple[str, Any, str]]
+
+    def __init__(self, **config):
+        base.PaddingMixin.__init__(self, **config)
+        self.add_defaults(_Decoration.defaults)
+
+    @property
+    def height(self):
+        return self.parent.height
+
+    @property
+    def width(self):
+        return self.parent.width
+
+    @property
+    def drawer(self):
+        return self.parent.drawer
+
+    @property
+    def ctx(self):
+        return self.parent.drawer.ctx
+
+    def draw(self, parent):
+        self.parent = parent
+
+
+class RectDecoration(_Decoration):
+    """
+        Widget decoration that draws a rectangle behind the widget contents.
+
+        Only one colour can be set but decorations can be layered to achieve
+        multi-coloured effects.
+
+        Rectangles can be drawn as just the the outline or filled. Curved corners
+        can be obtained by setting the ``radius`` parameter.
+    """
+    defaults = [
+        ("filled", False, "Whether to fill shape"),
+        ("radius", 4, "Radius for corners (0 for square)"),
+        ("dec_colour", "#000000", "Colour for decoration"),
+        ("linewidth", 2, "Line width for decoration")
+    ]  # type: List[Tuple[str, Any, str]]
+
+    def __init__(self, **config):
+        _Decoration.__init__(self, **config)
+        self.add_defaults(RectDecoration.defaults)
+
+    def draw(self, parent):
+        _Decoration.draw(self, parent)
+
+        box_height = self.height - 2 * self.padding_y
+        box_width = self.width - 2 * self.padding_x
+
+        self.drawer.set_source_rgb(self.dec_colour)
+
+        degrees = math.pi / 180.0
+
+        self.ctx.new_sub_path()
+
+        delta = self.radius + self.linewidth / 2
```

Comment:
Could this be optimised in the case that radius is 0, and just draw a rectangle directly?

Reply:
Yes.